File: go/src/fpSimulation/fluorescence.go
package fpSimulation

import (
	"math"
)

// This currently only handles Kalpha and Kbeta lines
// characteristics : Z, Edge, Kb, Ka, epsilon Kb, epsilon Ka
// Sample Format : Z, Ci(wtF)
// Output Format: Z, keVbeta, keValpha, Ibeta, Ialpha

func calculatePeaks(scen *Scenario, prop *Properties) {
	var nEnergies, nElements int
	var Pfluor, Sfluor, Ifluor float64

	sampleConfig(scen, prop)

	nEnergies = len(scen.BeamI)
	nElements = len(scen.Sample)
	for k := 0; k < nEnergies; k++ {
		for i := 0; i < nElements; i++ {
			for pi := 0; pi < 2; pi++ {
				Pfluor = primaryFluor(k, i, pi, scen, prop)
				Sfluor = 0.0
				for j := 0; j < nElements; j++ {
					for pj := 0; pj < 2; pj++ {
						Sfluor = Sfluor + secondaryFluor(Pfluor, k, i, pi, j, pj, scen, prop)
					}
				}
				Ifluor = scen.BeamI[k][1] * (Pfluor + Sfluor) * 0.02
				scen.PeaksI[i][pi+3] = scen.PeaksI[i][pi+3] + Ifluor
			}
		}
	}
	return
}

func primaryFluor(k int, i int, pi int, scen *Scenario, prop *Properties) float64 {
	var zi int
	var Pfluor, Ci, angleCorr, keV, keVedgeI, keVlineI, epsilonI float64
	var tauIE, muE, muLineI float64

	Pfluor = 0.0
	zi = (int)(scen.Sample[i][0])
	Ci = scen.Sample[i][1]
	if zi <= scen.MaxZ {
		angleCorr = math.Sin(45.0 * (math.Pi / 180.0))
		keV = scen.MuSample[k][0]
		keVedgeI = prop.Char[zi-1][1]
		keVlineI = prop.Char[zi-1][2+pi]
		epsilonI = prop.Char[zi-1][4+pi]
		tauIE = prop.Tau[k][zi]
		muE = scen.MuSample[k][1] / angleCorr
		muLineI = scen.MuSample[rowFromE(keVlineI)][1] / angleCorr
		if keV > keVedgeI {
			Pfluor = (scen.GeomFactor / angleCorr) * Ci * epsilonI * (tauIE / (muE + muLineI))
		}
	}
	return Pfluor
}

// Note that Ji(lambda_j)/Ji(lambda) is treated as 1 because JK is energy independent
func secondaryFluor(Pfluor float64, k int, i int, pi int, j int, pj int, scen *Scenario, prop *Properties) float64 {
	var zi, zj int
	var Sfluor, Cj, angleCorr, keV, keVedgeI, keVlineI, keVedgeJ, keVlineJ, epsilonJ float64
	var tauIE, tauJE, tauIlineJ, muE, muLineI, muLineJ, L float64

	Sfluor = 0.0
	zi = (int)(scen.Sample[i][0])
	zj = (int)(scen.Sample[j][0])
	Cj = scen.Sample[j][1]
	if (zi <= scen.MaxZ) && (zj <= scen.MaxZ) {
		angleCorr = math.Sin(45.0 * (math.Pi / 180.0))
		keV = scen.MuSample[k][0]
		keVedgeI = prop.Char[zi-1][1]
		keVlineI = prop.Char[zi-1][2+pi]
		keVedgeJ = prop.Char[zj-1][1]
		keVlineJ = prop.Char[zj-1][2+pj]
		epsilonJ = prop.Char[zj-1][4+pj]
		tauIE = prop.Tau[k][zi]
		tauJE = prop.Tau[k][zj]
		tauIlineJ = prop.Tau[rowFromE(keVlineJ)][zi]
		muE = scen.MuSample[k][1] / angleCorr
		muLineI = scen.MuSample[rowFromE(keVlineI)][1] / angleCorr
		muLineJ = scen.MuSample[rowFromE(keVlineJ)][1] // No angle correction for this term!
		if (keV > keVedgeI) && (keV > keVedgeJ) && (keVlineJ > keVedgeI) {
			L = (muLineJ/muLineI)*math.Log(1.0+(muLineI/muLineJ)) + (muLineJ/muE)*math.Log(1.0+(muE/muLineJ))
			Sfluor = 0.5 * Pfluor * Cj * epsilonJ * ((tauJE * tauIlineJ) / (tauIE * muLineJ)) * L
		}
	}
	return Sfluor
}

File: java/xrayCalculations/Fluorescence.java
package fpSimulation;

public class Fluorescence {
    double geomFactor;
    double angleCorrection;
    
    public Fluorescence(double geomFactor) {
        this.geomFactor = geomFactor;
        this.angleCorrection = Math.sin(Math.toRadians(45));
    }
    
    // This currently only handles Kalpha and Kbeta lines
    // characteristics : Z, Edge, Kb, Ka, epsilon Kb, epsilon Ka
    // Sample Format : Z, Ci(wtF)
    // Output Format: Z, keVbeta, keValpha, Ibeta, Ialpha
    
    public double[][] calculateSpectrum(double[][] beamI, double[][]sample, double[][] characteristics, double[][] tau, double[][] mu) {
        int nEnergies = beamI.length;
        int nElements = sample.length;
        double[][] peakI = createPeakArray(sample, characteristics);
        double[][] muSample = createMuSample(beamI, sample, mu);
        
        int zi;
        int zj;
        double Ci;
        double Cj;
        double Pfluor;
        double Sfluor;
        double Ifluor;
        for (int k = 0; k < nEnergies; k++) {
            for (int i = 0; i < nElements; i++) {
                zi = (int) sample[i][0];
                Ci = sample[i][1];
                for (int pi = 0; pi < 2; pi++) {
                    Pfluor = primaryFluor(k, zi, Ci, pi, characteristics, muSample, tau);
                    Sfluor = 0.0;
                    for (int j = 0; j < nElements; j++) {
                        zj = (int) sample[j][0];
                        Cj = sample[j][1];
                        for (int pj = 0; pj < 2; pj++) {
                            Sfluor = Sfluor + secondaryFluor(k, Pfluor, zi, zj, Cj, pi, pj, characteristics, muSample, tau);
                        }
                    }
                    Ifluor = beamI[k][1] * (Pfluor + Sfluor) * 0.02;
                    peakI[i][pi+3] = peakI[i][pi+3] + Ifluor;
                }
            }
        }
        
        return peakI;
    }
    
    private int rowFromE(double keV) {
        return (int) (keV/0.02);
    }
    
    private double[][] createPeakArray(double[][] sample, double[][] characteristics) {
        int nElements = sample.length;
        double[][] peakI = new double[nElements][5];
        int zi;
        for (int i = 0; i < nElements; i++) {
            zi = (int) sample[i][0];
            peakI[i][0] = sample[i][0];
            peakI[i][1] = characteristics[zi-1][2];
            peakI[i][2] = characteristics[zi-1][3];
            peakI[i][3] = 0.0;
            peakI[i][4] = 0.0;
        }
        return peakI;
    }
    
    private double[][] createMuSample(double[][] beamI, double[][] sample, double[][] mu) {
        int nEnergies = beamI.length;
        int nElements = sample.length;
        double[][] muSample = new double[nEnergies][2];
        int zi;
        double Ci;
        for (int k = 0; k < nEnergies; k++) {
            muSample[k][0] = beamI[k][0];
            for (int i = 0; i < nElements; i++) {
                zi = (int) sample[i][0];
                Ci = sample[i][1];
                muSample[k][1] = muSample[k][1] + (Ci * mu[k][zi]);
            }
        }
        return muSample;
    }
    
    private double primaryFluor(int k, int zi, double Ci, int pi, double[][] characteristics, double[][] muSample, double[][] tau) {
        double Pfluor = 0.0;
        double keV = muSample[k][0];
        double keVedgeI = characteristics[zi-1][1];
        double keVlineI = characteristics[zi-1][2+pi];
        double epsilonI = characteristics[zi-1][4+pi];
        double tauIE = tau[k][zi];
        double muE = muSample[k][1] / this.angleCorrection;
        double muLineI = muSample[rowFromE(keVlineI)][1] / this.angleCorrection;
        if (keV > keVedgeI) {
            Pfluor = (this.geomFactor/this.angleCorrection) * Ci * epsilonI * (tauIE / (muE + muLineI));
        }
        return Pfluor;
    }
    
    // Note that Ji(lambda_j)/Ji(lambda) is treated as 1 because JK is energy independent
    private double secondaryFluor(int k, double Pfluor, int zi, int zj, double Cj, int pi, int pj, double[][] characteristics, double[][] muSample, double[][] tau) {
        double Sfluor = 0.0;
        double keV = muSample[k][0];
        double keVedgeI = characteristics[zi-1][1];
        double keVlineI = characteristics[zi-1][2+pi];
        double epsilonI = characteristics[zi-1][4+pi];
        double keVedgeJ = characteristics[zj-1][1];
        double keVlineJ = characteristics[zj-1][2+pj];
        double epsilonJ = characteristics[zj-1][4+pj];
        double tauIE = tau[k][zi];
        double tauJE = tau[k][zj];
        double tauIlineJ = tau[rowFromE(keVlineJ)][zi];
        double muE = muSample[k][1] / this.angleCorrection;
        double muLineI = muSample[rowFromE(keVlineI)][1] / this.angleCorrection;
        double muLineJ = muSample[rowFromE(keVlineJ)][1];   // No angle correction for this term!
        if ( (keV > keVedgeI) && (keV > keVedgeJ) && (keVlineJ > keVedgeI) ) {
            double L = (muLineJ / muLineI) * Math.log(1.0 + (muLineI / muLineJ)) + (muLineJ / muE) * Math.log(1.0 + (muE / muLineJ));
            Sfluor = 0.5 * Pfluor * Cj * epsilonJ * ((tauJE * tauIlineJ)/(tauIE * muLineJ)) * L;
        }
        return Sfluor;
    }
}